### ISSUE TITLE
Cars to be able to use bus lanes to turn

### ISSUE THREAD

**Issue by kianzarrin** [CitiesSkylinesMods/TMPE#716]
![Screenshot (565)](https://user-images.githubusercontent.com/26344691/74840150-25be0580-532f-11ea-9271-06b7d8a8d94e.png)

### Situation:
In the real world cars can use bus lane to turn right. 
In the game they can't. Instead the bus comes into the car lane to go straight to avoid collisions (see screenshot above) - for a 12 lane road it works fine but if you want dedicated turning lanes on a 3 lane road then that's a problem. 

People are finding it difficult to use dedicated turning lanes feature on roads with bus lane. The timed traffic light quick-setup feature is also going to struggle for the same reason.

### Need
One way to solve this problem is to disable bus lane on the last segment.

but another more elegant solution is to somehow allow cars to use bus lanes only for turning.

### Action
I suggest to have a policy to allow cars to use bus lane only for turning (on the last segment only - and if its too short then its the user's problem!  ... it's better than nothing). 

adding such a policy would most certainly awaken an sleeping giant! Lane arrow and lane connector tools (or Routing tool #41) need to differentiate between car VS bus. 

![Screenshot (566)](https://user-images.githubusercontent.com/26344691/74843488-a7646200-5334-11ea-8b54-45c7ef998b80.png)

Lane arrow tool has GUI so we can just add more buttons there:
- bus lanes should have yellow GUI arrows. (regardless of the allow turn on bus lane policy).
- if the "allow turn on bus lane policy" is active then we need one more button to the side for car to turn on bus lane. it would be white color and the prop would also be white.
- In Israel they use yellow road arrow for bus directions and white road arrow for cars to use the bus lane to turn. I don't know about other countries. whatever the convention, we may need to patch game functions to implement bus+car props. This is important to visualize to user what is going on.
- bus lanes in the middle of the road would be unaffected by the policy.

For lane connector tool we need to add more control (modifier? hotkey? state machine? GUI) to be able to differentiate between bus and car lane.

This can also help auto dedicated turning lanes  see [BIFFA video at 23:47](https://youtu.be/JHFC89eVu6A?t=1427). but for now I think I should disable it on bus lane with a guide message asking people not to use bus lanes on junctions. Timed traffic light is going to be even more challenging. We need separate issue for those.

### Plan
~~This is EPIC. We need to break up the task:
Task 1: on the lane connector tool use yellow arrows for bus lanes (independent).
Task 2: use yellow props (or any other bus props) on base lanes (independent)
Task 3: dedicated turning lane policy. ( depends on task 1)
Task 4: Use advanced car+bus arrow props on bus lane(depends on task 2 and 3)
task 5: lane arrow connector to allow bus/car to connect separately (depends on 3).
EDIT: task 6: develop props/visual effects for countries that are different (this one is mostly for asset creators but maybe we need to provide some back end code for support)~~

from https://github.com/CitiesSkylinesMods/TMPE/issues/716#issuecomment-592252788 : 
> * We should allow users to override default (as set in mod options) vehicle restrictions on bus lanes
* We should add a new indented option in Policies tab:
    > [x] Ban ~~private~~ cars and trucks on bus lanes
    > ⠀⠀⠀[x] **Allow bus lanes near junctions to be used as turning lanes**
* The new option would auto-remove the vehicle restrictions from bus lanes on segments at junctions (user could override on lane by lane basis if they want to)
* Tool such as lane arrows, separated turning lanes, traffic lights, etc., could then adapt based on whether the bus lanes have the vehicle restriction or not.

**Comment by aubergine10** [comment]
I think we need more thought on this; it's quite different depending on country.

For example, in UK if I move in to a bus lane to make a turn it's instant fine time and possibly penalty points on my license... unless there is specific signage and road markings indicating that I am allowed to use that lane (effectively, the bus lane is temporarily put on hold until after the junction).

We definitely don't want to start telling users what roads to use at junctions, because different countries/regions have their own aproach to how bus lanes work.

**Comment by kianzarrin** [comment]
> We definitely don't want to start telling users what roads to use at junctions, because different countries/regions have their own approach to how bus lanes work.

But until we address this ticket there is no other choice. 

>  it's quite different depending on country.  

In Israel the law is not enforced lol! its a bit of a mess here and no one couldn't careless about something as meager as this when people (including police ... specially police) drive CRAZY! Its so funny in many cases lane switching is like skylines! People came here from all over the world and they brought their culture with them. Its much better than Iran though. At least zebra crossing means something.

But seriously ... I always see special road markings on bus lanes that for cars who want to turn right.

> unless there is specific signage and road markings indicating that I am allowed to use that lane 

Pay attention! that's exactly what I am saying. People can use traffic manager to setup rules on the bus lane. The "specific signage" would be yellow VS white lane arrow props. The prop could be customizable depending on the country. For UK its a bit hard:
![](https://www.drivingtesttips.biz/images/bus-lane-left-turn.jpg)

These are all visual effects though. The rules are the same in all countries. If there is a marking, cars can use the bus lane to turn.

I don't think we should go for all or nothing approach. First we support visual effects for 1 country, then we add visual effects for other countries too.

**Comment by aubergine10** [comment]
Decals for UK roads are in production :) https://github.com/aubergine10/uk_roads/issues/13

Vanilla game, and also default TMPE settings, allow cars/trucks to use bus lanes but deters them with a small pathfinder penalty.

TMPE has **Ban private cars and trucks on bus lanes** setting (Policies tab in mod options) which adds a more strict vehicle restriction (I think?) for cars/trucks in bus lanes. However, once enabled, there is no way for user to override that setting on the map.

So I think lane arrows is a bad place to achieve this feature, because what we're dealing with is essentially vehicel restrictions.

Suggestion:

* We should allow users to override default (as set in mod options) vehicle restrictions on bus lanes
* We should add a new indented option in Policies tab:
    > [x] Ban ~~private~~ cars and trucks on bus lanes
    > ⠀⠀⠀[x] **Allow bus lanes near junctions to be used as turning lanes**
* The new option would auto-remove the vehicle restrictions from bus lanes on segments at junctions (user could override on lane by lane basis if they want to)
* Tool such as lane arrows, separated turning lanes, traffic lights, etc., could then adapt based on whether the bus lanes have the vehicle restriction or not.

If user wants something more visually elaborate, they can upgrade the segments at junctions to a road asset that doesn't have bus lanes and then add their own country-specific decals or whatever.

In regards to having TMPE add the visual effects, it would be nice but I think it is probably best left until a later date.

**Comment by kianzarrin** [comment]
@aubergine10 Cool stuff! I didn't know that.

> If user wants something more visually elaborate, they can upgrade the segments at junctions to a road asset that doesn't have bus lanes and then add their own country-specific decals or whatever.

who knows? maybe I make a mod for this! You are right though its mostly visual effects.

**Comment by aubergine10** [comment]
Just when we thought it was safe to let cars use bus lanes at junctions, this appears: [Guided busways](https://steamcommunity.com/sharedfiles/filedetails/?id=2003910724)

![image](https://user-images.githubusercontent.com/1386719/75505165-6a523c80-59d2-11ea-982b-43ee7e2e89ea.png)

It seems the 'benefits' of these things are:

* Narrower than normal roads = takes up less space
* Requires less construction materials
* Busses can drive faster as they are 'guided' via optical/mechanical/magnetic "tracks"

Cars/trucks entering these things would be bad, for example a [lorry got stuck in one for 6 hours in UK](https://en.wikipedia.org/wiki/Cambridgeshire_Guided_Busway).

**Comment by kianzarrin** [comment]
AR mod should take care of this
